pkg/endpoint/policy.go

```diff
@@ -309,16 +309,28 @@ func (e *Endpoint) regeneratePolicy(owner Owner) (bool, error) {
 	// already running the latest revision, otherwise we have to wait for
 	// the regeneration of the endpoint to complete.
 	if !policyChanged {
-		e.PolicyRevision = revision
+		e.policyRevision = revision
 	} else {
-		e.NextPolicyRevision = revision
+		e.nextPolicyRevision = revision
 	}
 
 	return policyChanged || optsChanged, nil
 }
 
 // Called with e.Mutex locked
 func (e *Endpoint) regenerate(owner Owner) error {
+	e.BuildMutex.Lock()
+	defer e.BuildMutex.Unlock()
+
+	// If endpoint was marked as disconnected then
+	// it won't be regenerated
+	if e.IsDisconnecting() {
+		log.Debugf("Endpoint %d disconnected, skipping build", e.ID)
+		return fmt.Errorf("endpoint disconnected, skipping build")
+	}
+
+	log.Debugf("Regenerating endpoint %d...", e.ID)
```

Comment:
Can we have logs with the endpoint ID use log.WIthFields? E.g.:
```
log.WithFields( log.Fields{
                                logfields.EndpointID: ep.StringID(), ...
```

Reply:
Fixed